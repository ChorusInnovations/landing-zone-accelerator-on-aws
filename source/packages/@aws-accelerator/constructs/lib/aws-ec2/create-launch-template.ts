/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as fs from 'fs';

export interface ILaunchTemplateResource extends cdk.IResource {
  /**
   * The version number.
   */
  readonly version: string;
  /**
   * The ID of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateId: string;
  /**
   * The name of the AWS::EC2::LaunchTemplate . You must specify either a LaunchTemplateName or a LaunchTemplateId .
   */
  readonly launchTemplateName: string;
}
export type PrivateIpAddressConfig = {
  primary: boolean | undefined;
  privateIpAddress: string | undefined;
};
export type NetworkInterfaceItemConfig = {
  associateCarrierIpAddress: boolean | undefined;
  associatePublicIpAddress: boolean | undefined;
  deleteOnTermination: boolean | undefined;
  description: string | undefined;
  deviceIndex: number | undefined;
  groups: string[] | undefined;
  interfaceType: string | undefined;
  networkCardIndex: number | undefined;
  networkInterfaceId: string | undefined;
  privateIpAddress: string | undefined;
  secondaryPrivateIpAddressCount: number | undefined;
  subnetId: string | undefined;
  privateIpAddresses: PrivateIpAddressConfig[] | undefined;
};
export type EbsProperty = {
  deleteOnTermination?: boolean;
  encrypted?: boolean;
  iops?: number;
  kmsKeyId?: string;
  snapshotId?: string;
  throughput?: number;
  volumeSize?: number;
  volumeType?: string;
};

export type BlockDeviceMappingItem = {
  deviceName: string;
  ebs?: EbsProperty;
};

export interface LaunchTemplateProps {
  /*
   * Path to user data.
   */
  readonly userData?: string;
  /*
   * VpcName
   */
  readonly vpc: string;
  /*
   * Name of Launch Template
   */
  readonly name: string;
  /*
   * Name of Application
   */
  readonly appName: string;
  readonly blockDeviceMappings?: BlockDeviceMappingItem[];
  readonly securityGroups?: string[];
  readonly networkInterfaces?: NetworkInterfaceItemConfig[];
  readonly instanceType: string;
  readonly keyPair?: string;
  readonly iamInstanceProfile?: string;
  readonly imageId: string;
  readonly enforceImdsv2?: boolean;
}

export class LaunchTemplate extends cdk.Resource implements ILaunchTemplateResource {
  public readonly version: string;
  public readonly launchTemplateId: string;
  public readonly launchTemplateName: string;

  constructor(scope: Construct, id: string, props: LaunchTemplateProps) {
    super(scope, id);

    let metadataOptions: cdk.aws_ec2.CfnLaunchTemplate.MetadataOptionsProperty;

    if (props.enforceImdsv2 === false) {
      metadataOptions = { httpTokens: 'optional' };
    } else {
      metadataOptions = { httpTokens: 'required' };
    }

    const launchTemplate = new cdk.aws_ec2.CfnLaunchTemplate(this, 'LaunchTemplate', {
      launchTemplateData: {
        blockDeviceMappings: props.blockDeviceMappings ?? undefined,
        securityGroupIds: props.securityGroups ?? undefined,
        networkInterfaces: props.networkInterfaces ?? undefined,
        instanceType: props.instanceType,
        keyName: props.keyPair ?? undefined,
        imageId: props.imageId,
        iamInstanceProfile: { name: props.iamInstanceProfile ?? undefined },
        userData: cdk.Fn.base64(fs.readFileSync(props.userData!, 'utf8')) ?? undefined,
        metadataOptions,
      },
      launchTemplateName: props.name,
    });

    this.launchTemplateId = launchTemplate.ref;
    this.launchTemplateName = launchTemplate.launchTemplateName!;
    this.version = launchTemplate.attrLatestVersionNumber;
  }
}
