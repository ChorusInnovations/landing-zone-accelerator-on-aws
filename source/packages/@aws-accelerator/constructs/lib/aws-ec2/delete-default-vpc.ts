/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { v4 as uuidv4 } from 'uuid';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Class Delete the Default VPC
 */
export class DeleteDefaultVpc extends Construct {
  readonly id: string;
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const DELETE_DEFAULT_VPC_TYPE = 'Custom::DeleteDefaultVpc';

    //
    // Function definition for the custom resource
    //
    const cr = cdk.CustomResourceProvider.getOrCreateProvider(this, DELETE_DEFAULT_VPC_TYPE, {
      codeDirectory: path.join(__dirname, 'delete-default-vpc/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_14_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: [
            'ec2:DeleteInternetGateway',
            'ec2:DetachInternetGateway',
            'ec2:DeleteNetworkAcl',
            'ec2:DeleteRoute',
            'ec2:DeleteSecurityGroup',
            'ec2:DeleteSubnet',
            'ec2:DeleteVpc',
            'ec2:DescribeInternetGateways',
            'ec2:DescribeNetworkAcls',
            'ec2:DescribeRouteTables',
            'ec2:DescribeSecurityGroups',
            'ec2:DescribeSubnets',
            'ec2:DescribeVpcs',
          ],
          Resource: '*',
        },
      ],
    });

    //
    // Custom Resource definition. We want this resource to be evaluated on
    // every CloudFormation update, so we generate a new uuid to force
    // re-evaluation.
    //
    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: DELETE_DEFAULT_VPC_TYPE,
      serviceToken: cr.serviceToken,
      properties: {
        uuid: uuidv4(),
      },
    });

    this.id = resource.ref;
  }
}
