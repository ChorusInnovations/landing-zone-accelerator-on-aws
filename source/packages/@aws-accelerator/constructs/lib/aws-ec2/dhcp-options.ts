/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnTag, IResource, Resource } from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface IDhcpOptions extends IResource {
  /**
   * The name of the DHCP options set.
   */
  readonly name: string;

  /**
   * The ID of the DHCP options set.
   */
  readonly dhcpOptionsId: string;
}

export interface DhcpOptionsProps {
  /**
   * The name of the DHCP options set.
   */
  readonly name: string;

  /**
   * This value is used to complete unqualified DNS hostnames.
   */
  readonly domainName?: string;

  /**
   * The IPv4 addresses of up to four domain name servers.
   *
   * @default -- AmazonProvidedDNS
   */
  readonly domainNameServers?: string[];

  /**
   * The IPv4 addresses of up to four NetBIOS name servers.
   */
  readonly netbiosNameServers?: string[];

  /**
   * The NetBIOS node type (1, 2, 4, or 8).
   */
  readonly netbiosNodeType?: number;

  /**
   * The IPv4 addresses of up to four Network Time Protocol (NTP) servers.
   */
  readonly ntpServers?: string[];

  /**
   * Any tags assigned to the DHCP options set.
   */
  readonly tags?: CfnTag[];
}

export class DhcpOptions extends Resource implements IDhcpOptions {
  public readonly name: string;
  public readonly dhcpOptionsId: string;

  constructor(scope: Construct, id: string, props: DhcpOptionsProps) {
    super(scope, id);

    this.name = props.name;

    // Add name tag to tags
    props.tags?.push({ key: 'Name', value: this.name });

    const resource = new ec2.CfnDHCPOptions(this, 'Resource', {
      domainName: props.domainName,
      domainNameServers: props.domainNameServers ?? ['AmazonProvidedDNS'],
      netbiosNameServers: props.netbiosNameServers,
      netbiosNodeType: props.netbiosNodeType,
      ntpServers: props.ntpServers,
      tags: props.tags,
    });

    this.dhcpOptionsId = resource.attrDhcpOptionsId;
  }
}
