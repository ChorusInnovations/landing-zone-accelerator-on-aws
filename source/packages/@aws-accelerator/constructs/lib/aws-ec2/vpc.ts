/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as core from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface IRouteTable extends core.IResource {
  /**
   * The identifier of the route table
   *
   * @attribute
   */
  readonly routeTableId: string;

  /**
   * The VPC associated with the route table
   *
   * @attribute
   */
  readonly vpc: IVpc;
}

export interface RouteTableProps {
  readonly name: string;
  readonly vpc: IVpc;
}

export class RouteTable extends core.Resource implements IRouteTable {
  public readonly routeTableId: string;

  public readonly vpc: IVpc;

  constructor(scope: Construct, id: string, props: RouteTableProps) {
    super(scope, id);

    this.vpc = props.vpc;

    const resource = new ec2.CfnRouteTable(this, 'Resource', {
      vpcId: props.vpc.vpcId,
      tags: [{ key: 'Name', value: props.name }],
    });

    this.routeTableId = resource.ref;
  }

  public addTransitGatewayRoute(
    id: string,
    destination: string,
    transitGatewayId: string,
    transitGatewayAttachment: core.CfnResource,
  ): void {
    const route = new ec2.CfnRoute(this, id, {
      routeTableId: this.routeTableId,
      destinationCidrBlock: destination,
      transitGatewayId: transitGatewayId,
    });
    route.addDependsOn(transitGatewayAttachment);
  }

  public addNatGatewayRoute(id: string, destination: string, natGatewayId: string): void {
    new ec2.CfnRoute(this, id, {
      routeTableId: this.routeTableId,
      destinationCidrBlock: destination,
      natGatewayId: natGatewayId,
    });
  }

  public addInternetGatewayRoute(id: string, destination: string): void {
    if (!this.vpc.internetGatewayId) {
      throw new Error('Attempting to add Internet Gateway route without an IGW defined.');
    }

    new ec2.CfnRoute(this, id, {
      routeTableId: this.routeTableId,
      destinationCidrBlock: destination,
      gatewayId: this.vpc.internetGatewayId,
    });
  }
}

export interface ISubnet extends core.IResource {
  /**
   * The identifier of the subnet
   *
   * @attribute
   */
  readonly subnetId: string;

  /**
   * The name of the subnet
   *
   * @attribute
   */
  readonly subnetName: string;

  /**
   * The Availability Zone the subnet is located in
   *
   * @attribute
   */
  readonly availabilityZone: string;

  // /**
  //  * The VPC associated with the subnet
  //  *
  //  * @attribute
  //  */
  // readonly routeTable: IRouteTable;

  //  /**
  //   * The route table for this subnet
  //   */
  //  readonly routeTable: IRouteTable;

  //  /**
  //   * Associate a Network ACL with this subnet
  //   *
  //   * @param acl The Network ACL to associate
  //   */
  //  associateNetworkAcl(id: string, acl: INetworkAcl): void;

  // /**
  //  * The IPv4 CIDR block for this subnet
  //  */
  //  readonly ipv4CidrBlock: string;

  //  readonly mapPublicIpOnLaunch: boolean;
}

export interface SubnetProps {
  readonly name: string;
  readonly availabilityZone: string;
  readonly ipv4CidrBlock: string;
  readonly mapPublicIpOnLaunch?: boolean;
  readonly routeTable: IRouteTable;
  readonly vpc: IVpc;
  // readonly nacl: INacl;
}

export class Subnet extends core.Resource implements ISubnet {
  public readonly subnetName: string;
  public readonly availabilityZone: string;
  public readonly ipv4CidrBlock: string;
  public readonly mapPublicIpOnLaunch?: boolean;
  public readonly routeTable: IRouteTable;
  public readonly subnetId: string;

  constructor(scope: Construct, id: string, props: SubnetProps) {
    super(scope, id);

    this.subnetName = props.name;
    this.availabilityZone = props.availabilityZone;
    this.ipv4CidrBlock = props.ipv4CidrBlock;
    this.mapPublicIpOnLaunch = props.mapPublicIpOnLaunch;
    this.routeTable = props.routeTable;

    const resource = new ec2.CfnSubnet(this, 'Resource', {
      vpcId: props.vpc.vpcId,
      cidrBlock: props.ipv4CidrBlock,
      availabilityZone: props.availabilityZone,
      mapPublicIpOnLaunch: props.mapPublicIpOnLaunch,
      tags: [{ key: 'Name', value: props.name }],
    });

    this.subnetId = resource.ref;

    new ec2.CfnSubnetRouteTableAssociation(this, 'RouteTableAssociation', {
      subnetId: this.subnetId,
      routeTableId: props.routeTable.routeTableId,
    });
  }
}

export interface INatGateway extends core.IResource {
  /**
   * The identifier of the NAT Gateway
   *
   * @attribute
   */
  readonly natGatewayId: string;

  /**
   * The name of the NAT Gateway
   *
   * @attribute
   */
  readonly natGatewayName: string;
}

export interface NatGatewayProps {
  readonly name: string;
  readonly subnet: ISubnet;
}

export class NatGateway extends core.Resource implements INatGateway {
  public readonly natGatewayId: string;
  public readonly natGatewayName: string;

  constructor(scope: Construct, id: string, props: NatGatewayProps) {
    super(scope, id);

    this.natGatewayName = props.name;

    const resource = new ec2.CfnNatGateway(this, 'Resource', {
      subnetId: props.subnet.subnetId,
      allocationId: new ec2.CfnEIP(this, 'Eip', {
        domain: 'vpc',
      }).attrAllocationId,
      tags: [{ key: 'Name', value: props.name }],
    });

    this.natGatewayId = resource.ref;
  }
}

export interface IVpc extends core.IResource {
  /**
   * The identifier of the vpc
   *
   * @attribute
   */
  readonly vpcId: string;

  /**
   * The Internet Gateway Id
   */
  readonly internetGatewayId?: string;
}

/**
 * Construction properties for a  VPC object.
 */
export interface VpcProps {
  readonly name: string;
  readonly ipv4CidrBlock: string;
  readonly enableDnsHostnames?: boolean;
  readonly enableDnsSupport?: boolean;
  readonly instanceTenancy?: 'default' | 'dedicated';
  readonly internetGateway?: boolean;
}

/**
 * Defines a  S3 Bucket object. By default a KMS CMK is generated and
 * associated to the bucket.
 */
export class Vpc extends core.Resource implements IVpc {
  // private readonly vpc: ec2.CfnVPC;

  public readonly vpcId: string;
  public readonly internetGatewayId: string | undefined;

  constructor(scope: Construct, id: string, props: VpcProps) {
    super(scope, id);

    const resource = new ec2.CfnVPC(this, 'Resource', {
      cidrBlock: props.ipv4CidrBlock,
      enableDnsHostnames: props.enableDnsHostnames,
      enableDnsSupport: props.enableDnsSupport,
      instanceTenancy: props.instanceTenancy,
      tags: [{ key: 'Name', value: props.name }],
    });

    this.vpcId = resource.ref;

    if (props.internetGateway) {
      const igw = new ec2.CfnInternetGateway(this, 'InternetGateway', {});

      new ec2.CfnVPCGatewayAttachment(this, 'InternetGatewayAttachment', {
        internetGatewayId: igw.ref,
        vpcId: this.vpcId,
      });

      this.internetGatewayId = igw.ref;
    }
  }

  public addGatewayVpcEndpoint(id: string, service: string, routeTableIds: string[]): void {
    // const gwService = ;
    new ec2.CfnVPCEndpoint(this, id, {
      serviceName: new ec2.GatewayVpcEndpointAwsService(service).name,
      vpcId: this.vpcId,
      routeTableIds,
    });
  }

  // public getVpc(): ec2.CfnVPC {
  //   return this.vpc;
  // }
}
