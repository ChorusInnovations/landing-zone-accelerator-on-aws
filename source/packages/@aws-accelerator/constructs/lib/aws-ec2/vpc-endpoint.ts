/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { IVpc, ISubnet, ISecurityGroup } from './vpc';
import { IRouteTable } from './route-table';

export interface IVpcEndpoint extends cdk.IResource {
  readonly vpcEndpointId: string;
  readonly service: string;
  readonly vpc: IVpc;
  readonly dnsName?: string;
  readonly hostedZoneId?: string;
}

export interface VpcEndpointProps {
  readonly vpc: IVpc;
  readonly vpcEndpointType: cdk.aws_ec2.VpcEndpointType;
  readonly service: string;
  readonly subnets?: ISubnet[];
  readonly securityGroups?: ISecurityGroup[];
  readonly privateDnsEnabled?: boolean;
  readonly policyDocument?: cdk.aws_iam.PolicyDocument;
  readonly routeTables?: IRouteTable[];
}

export class VpcEndpoint extends cdk.Resource implements IVpcEndpoint {
  public readonly vpcEndpointId: string;

  public readonly vpc: IVpc;
  public readonly service: string;
  public readonly dnsName?: string;
  public readonly hostedZoneId?: string;

  constructor(scope: Construct, id: string, props: VpcEndpointProps) {
    super(scope, id);

    this.vpc = props.vpc;
    this.service = props.service;

    if (props.vpcEndpointType === cdk.aws_ec2.VpcEndpointType.INTERFACE) {
      let serviceName = `com.amazonaws.${cdk.Stack.of(this).region}.${props.service}`;
      if (props.service in ['notebook', 'studio']) {
        serviceName = `aws.sagemaker.${cdk.Stack.of(this).region}.${props.service}`;
      }
      if (props.service in ['s3-global.accesspoint']) {
        serviceName = `com.aws.${props.service}`;
      }

      const resource = new cdk.aws_ec2.CfnVPCEndpoint(this, 'Resource', {
        serviceName,
        vpcEndpointType: props.vpcEndpointType,
        vpcId: props.vpc.vpcId,
        subnetIds: props.subnets?.map(item => item.subnetId),
        securityGroupIds: props.securityGroups?.map(item => item.securityGroupId),
        privateDnsEnabled: props.privateDnsEnabled,
        policyDocument: props.policyDocument,
      });
      this.vpcEndpointId = resource.ref;

      let dnsEntriesIndex = 0;
      if (props.service in ['notebook', 'studio']) {
        // TODO Top 3 DNS names are not valid so selecting the 4th DNS
        // need to find a better way to identify the valid DNS for PHZ
        dnsEntriesIndex = 4;
      }

      this.dnsName = cdk.Fn.select(1, cdk.Fn.split(':', cdk.Fn.select(dnsEntriesIndex, resource.attrDnsEntries)));
      this.hostedZoneId = cdk.Fn.select(0, cdk.Fn.split(':', cdk.Fn.select(dnsEntriesIndex, resource.attrDnsEntries)));
      return;
    }

    if (props.vpcEndpointType === cdk.aws_ec2.VpcEndpointType.GATEWAY) {
      const resource = new cdk.aws_ec2.CfnVPCEndpoint(this, 'Resource', {
        serviceName: new cdk.aws_ec2.GatewayVpcEndpointAwsService(props.service).name,
        vpcId: props.vpc.vpcId,
        routeTableIds: props.routeTables?.map(item => item.routeTableId),
        policyDocument: props.policyDocument,
      });
      this.vpcEndpointId = resource.ref;
      return;
    }

    throw new Error('Invalid vpcEndpointType specified');
  }
}
