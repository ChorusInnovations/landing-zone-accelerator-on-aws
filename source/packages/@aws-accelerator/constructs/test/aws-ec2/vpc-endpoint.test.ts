/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';

import { SynthUtils } from '@aws-cdk/assert';

import { SecurityGroup } from '../../lib/aws-ec2/vpc';
import { VpcEndpoint, VpcEndpointType } from '../../lib/aws-ec2/vpc-endpoint';

const testNamePrefix = 'Construct(VpcEndpoint): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

const securityGroup = new SecurityGroup(stack, 'TestSecurityGroup`', {
  securityGroupName: 'TestSecurityGroup',
  description: `AWS Private Endpoint Zone`,
  vpcId: 'Test',
});

new VpcEndpoint(stack, 'VpcEndpoint', {
  vpcId: 'Test',
  vpcEndpointType: VpcEndpointType.GATEWAY,
  service: 'service',
  subnets: ['Test1', 'Test2'],
  securityGroups: [securityGroup],
  privateDnsEnabled: true,
  policyDocument: new cdk.aws_iam.PolicyDocument({
    statements: [
      new cdk.aws_iam.PolicyStatement({
        sid: 'AccessToTrustedPrincipalsAndResources',
        actions: ['*'],
        effect: cdk.aws_iam.Effect.ALLOW,
        resources: ['*'],
        principals: [new cdk.aws_iam.AnyPrincipal()],
        conditions: {
          StringEquals: {
            'aws:PrincipalOrgID': ['organizationId'],
          },
        },
      }),
    ],
  }),
  routeTables: ['Test1', 'Test2'],
});

/**
 * VpcEndpoint construct test
 */
describe('VpcEndpoint', () => {
  /**
   * Snapshot test
   */
  test(`${testNamePrefix} Snapshot Test`, () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  /**
   * Number of SecurityGroup test
   */
  test(`${testNamePrefix} SecurityGroup resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::EC2::SecurityGroup', 1);
  });

  /**
   * Number of VPCEndpoint test
   */
  test(`${testNamePrefix} VPCEndpoint resource count test`, () => {
    cdk.assertions.Template.fromStack(stack).resourceCountIs('AWS::EC2::VPCEndpoint', 1);
  });

  /**
   * SecurityGroup resource configuration test
   */
  test(`${testNamePrefix} SecurityGroup resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        TestSecurityGroupDA4B5F83: {
          Type: 'AWS::EC2::SecurityGroup',
          Properties: {
            GroupDescription: 'AWS Private Endpoint Zone',
            GroupName: 'TestSecurityGroup',
            Tags: [
              {
                Key: 'Name',
                Value: 'TestSecurityGroup',
              },
            ],
            VpcId: 'Test',
          },
        },
      },
    });
  });

  /**
   * VPC endpoint resource configuration test
   */
  test(`${testNamePrefix} VPC endpoint resource configuration test`, () => {
    cdk.assertions.Template.fromStack(stack).templateMatches({
      Resources: {
        VpcEndpoint80208C18: {
          Type: 'AWS::EC2::VPCEndpoint',
          Properties: {
            PolicyDocument: {
              Statement: [
                {
                  Action: '*',
                  Condition: {
                    StringEquals: {
                      'aws:PrincipalOrgID': ['organizationId'],
                    },
                  },
                  Effect: 'Allow',
                  Principal: {
                    AWS: '*',
                  },
                  Resource: '*',
                  Sid: 'AccessToTrustedPrincipalsAndResources',
                },
              ],
              Version: '2012-10-17',
            },
            RouteTableIds: ['Test1', 'Test2'],
            ServiceName: {
              'Fn::Join': [
                '',
                [
                  'com.amazonaws.',
                  {
                    Ref: 'AWS::Region',
                  },
                  '.service',
                ],
              ],
            },
            VpcId: 'Test',
          },
        },
      },
    });
  });
});
