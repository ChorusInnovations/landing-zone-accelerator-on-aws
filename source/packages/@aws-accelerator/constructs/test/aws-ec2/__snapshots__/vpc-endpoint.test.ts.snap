// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VpcEndpoint Construct(VpcEndpoint):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "TestSecurityGroupDA4B5F83": Object {
      "Properties": Object {
        "GroupDescription": "AWS Private Endpoint Zone",
        "GroupName": "TestSecurityGroup",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestSecurityGroup",
          },
        ],
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcEndpoint80208C18": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": Array [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "RouteTableIds": Array [
          "Test1",
          "Test2",
        ],
        "ServiceName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "com.amazonaws.",
              Object {
                "Ref": "AWS::Region",
              },
              ".service",
            ],
          ],
        },
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceEc21C025642": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": Array [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": Array [
          Object {
            "Ref": "TestSecurityGroupDA4B5F83",
          },
        ],
        "ServiceName": "ec2",
        "SubnetIds": Array [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "Interface",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceGwlb3F654CE2": Object {
      "Properties": Object {
        "ServiceName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "com.amazonaws.vpce.",
              Object {
                "Ref": "AWS::Region",
              },
              ".ec2",
            ],
          ],
        },
        "SubnetIds": Array [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "GatewayLoadBalancer",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceS343EF5B95": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": Array [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": Array [
          Object {
            "Ref": "TestSecurityGroupDA4B5F83",
          },
        ],
        "ServiceName": "com.aws.s3-global.accesspoint",
        "SubnetIds": Array [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "Interface",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceSagemaker70CE2242": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": Array [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": Array [
          Object {
            "Ref": "TestSecurityGroupDA4B5F83",
          },
        ],
        "ServiceName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "aws.sagemaker.",
              Object {
                "Ref": "AWS::Region",
              },
              ".notebook",
            ],
          ],
        },
        "SubnetIds": Array [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "Interface",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointid71FBEB7E": Object {
      "Properties": Object {
        "DestinationCidrBlock": "10.100.0.0/16",
        "RouteTableId": "routeTableId",
        "VpcEndpointId": Object {
          "Ref": "VpcEndpoint80208C18",
        },
      },
      "Type": "AWS::EC2::Route",
    },
  },
}
`;
